browser/component_updater/brave_component_contents_verifier.cc

```diff
@@ -0,0 +1,193 @@
+    std::unique_ptr<crypto::SecureHash> hash(
+        crypto::SecureHash::Create(crypto::SecureHash::SHA256));
+    hash->Update(contents.subspan(offset, bytes_to_read));
+
+    std::string buffer;
+    buffer.resize(crypto::kSHA256Length);
+    hash->Finish(base::as_writable_byte_span(buffer));
+    hashes.push_back(std::move(buffer));
+
+    // If |contents| is empty, then we want to just exit here.
+    if (bytes_to_read == 0) {
+      break;
+    }
+
+    offset += bytes_to_read;
+  } while (offset < contents.size());
+
+  CHECK(block_size % crypto::kSHA256Length == 0);
+  return extensions::ComputeTreeHashRoot(hashes,
+                                         block_size / crypto::kSHA256Length);
+}
+
+// Proxies the file access and checks the verified_contents.json.
+class ComponentContentsAccessorImpl
+    : public ComponentNoChecksContentsAccessorImpl {
+ public:
+  explicit ComponentContentsAccessorImpl(const base::FilePath& component_root)
+      : ComponentNoChecksContentsAccessorImpl(component_root) {
+    verified_contents_ = extensions::VerifiedContents::CreateFromFile(
+        base::make_span(kComponentContentsVerifierPublicKey),
+        component_root.AppendASCII(kVerifiedContentsPath));
+    if (verified_contents_->block_size() % crypto::kSHA256Length != 0) {
+      // Unsupported block size.
+      verified_contents_.reset();
+    }
+  }
+
+  bool IsComponentSignatureValid() const override {
+    return !!verified_contents_.get();
+  }
+
+  void IgnoreInvalidSignature(bool ignore) override {
+    ignore_invalid_signature_ = ignore;
+  }
+
+  bool VerifyContents(const base::FilePath& relative_path,
+                      base::span<const uint8_t> contents) override {
+    if (!IsComponentSignatureValid()) {
+      return ignore_invalid_signature_;
+    }
+
+    if (!verified_contents_->HasTreeHashRoot(relative_path)) {
+      return true;
+    }
+
+    const auto root_hash =
+        GetRootHasheForContent(contents, verified_contents_->block_size());
+    if (!verified_contents_->TreeHashRootEquals(relative_path, root_hash)) {
+      return false;
+    }
+    return true;
+  }
+
+ private:
+  ~ComponentContentsAccessorImpl() override = default;
+
+  std::unique_ptr<extensions::VerifiedContents> verified_contents_;
+  bool ignore_invalid_signature_ = false;
+};
+
+}  // namespace
+
+#endif  // BUILDFLAG(ENABLE_EXTENSIONS)
+
+namespace component_updater {
+
+scoped_refptr<brave_component_updater::ComponentContentsAccessor>
+CreateComponentContentsAccessor(bool with_verifier,
+                                const base::FilePath& component_root) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  if (with_verifier) {
+    return base::MakeRefCounted<ComponentContentsAccessorImpl>(component_root);
+  }
+#endif
+  // if there is no extensions enabled then we expect that on these platforms
+  // the component files are protected by the OS.
+  return base::MakeRefCounted<ComponentNoChecksContentsAccessorImpl>(
+      component_root);
+}
```

Comment:
We have `ComponentContentsAccessor` ->  `ComponentNoChecksContentsAccessorImpl` -> `ComponentContentsAccessorImpl`, and then we wrap this all behind a repeating callback to `CreateComponentContentsAccessor`.

`ComponentContentsAccessorImpl` with `ignore_invalid_signature_=true` is close enough to `ComponentNoChecksContentsAccessorImpl`.

I'm wondering if subclassing things like this is really a gain. The main difference here is that certain build configurations don't have `BUILDFLAG(ENABLE_EXTENSIONS)`. I feel this should all be one single class, namely, `ComponentContentsAccessor`, and then a different `.cc` for build configs that don't support extension. This would avoid the whole use of inheritance here.

Reply:
Right, that makes sense. Thanks for clarifying. It seems to me then that this very specific code should be part of something like a delegate, rather than just be subclassing `ComponentContentsAccessor`, and that delegate should be passed into `ComponentContentsAccessor` however way it gets constructed.

Reply:
To clarify maybe this should be something like this:

```cxx
std::unique_ptr<brave_component_updater::ComponentContentsAccessor::Delegate>
CreateComponentContentsAccessor(bool with_verifier,
                                const base::FilePath& component_root) {
```

and

```cxx
void SetupComponentContentsVerifier() {
  auto factory = base::BindRepeating(CreateComponentContentsAccessor, true);
  brave_component_updater::ComponentContentsAccessor::Setup(std::move(factory));
}
```

What are your thoughts?